### PR TITLE
Purchase course run again

### LINKED ISSUE #3472: Allow learners to pay for a course run again if ran out of exam attempts
Might be solved by #3326

If learner ran out exam attempts, a course status message offers to pay for course again. Need to change the validation logic to allow for the course_run to be purchased again.

### PR DESCRIPTION
**Pull request by annagav**
#### What are the relevant tickets?
Fix #3472 

#### What's this PR do?
Let user go through the payment for a course run that was already payed in the past, when the user has the right amount of failed exam attempts.

#### How should this be manually tested?
Need to get to the state where you have paid for a courserun and have 2 failed exam attempts. Make sure the 2 exam grades are for 2 different exam runs. Then the course status message will ask you to pay and show a PAY button. You should be able to click CONTINUE on the Order Summary page and to actual payment details page.

Login as a learner enrolled in FA program like 'Digital Learning'. Run alter_data command for your user to have passed a course run:
`./manage.py alter_data set_past_run_to_passed --username username --course-title 'Digital Learning 100'`
Then in the django shell set up the exam:
```
from courses.models import *
from django.contrib.auth.models import User
from exams.factories import ExamRunFactory
from exams.models import *

c=Course.objects.filter(title='Digital Learning 100')[0]

exam_run=ExamRunFactory.create(course=c, authorized=True)

user = User.objects.filter(username='username')[0]

au=ExamAuthorizationFactory.create(user=user, course=c, exam_run=exam_run, status= 'success', exam_taken=True)

exam_profile= ExamProfileFactory.create(status='success', profile=user.profile)
```
Create failed exam grade
```
from grades.factories import *
from micromasters.utils import now_in_utc
from datetime import timedelta

exam_grade = ProctoredExamGradeFactory.create(user=user, course=c, exam_run=exam_run, passed=False, percentage_grade=0.1)
exam_run.date_grades_available=now_in_utc()-timedelta(days=1)
exam_run.save()
```

### CODE REVIEW

File: ecommerce/api.py [hunk 1]
```diff
@@ -96,8 +97,10 @@ def get_purchasable_course_run(course_key, user):
             order__user=user,
             course_key=course_run.edx_course_key,
     ).exists():
-        log.warning("Course run %s is already purchased by user %s", course_key, user)
-        raise ValidationError("Course run {} is already purchased".format(course_key))
+        mmtrack = get_mmtrack(user, course_run.course.program)
+        if not has_to_pay_for_exam(mmtrack, course_run.course):
+            log.warning("Course run %s is already purchased by user %s", course_key, user)
+            raise ValidationError("Course run {} is already purchased".format(course_key))
```
Comment (amir-qayyum-khan) [inline ecommerce/api.py]: please add a test to assert `ValidationError` raise

File: ecommerce/api.py [hunk 2]
```diff
@@ -96,8 +97,10 @@ def get_purchasable_course_run(course_key, user):
             order__user=user,
             course_key=course_run.edx_course_key,
```
Comment (amir-qayyum-khan) [inline ecommerce/api.py]: is this enough to figure out that user has paid for course?
i think these filters are unique for edx_course_key
```python
order__status=Order.FULFILLED,
order__user=user,
 course_key=course_run.edx_course_key,
```


```python
if Line.objects.filter(
            order__status=Order.FULFILLED,
            order__user=user,
            course_key=course_run.edx_course_key,
    ).exists():
```

Comment (amir-qayyum-khan) [inline ecommerce/api.py]: so for a course user can attempt 1 time exam in one payment, the for each new exam authorization he needs to  purchase the run

Comment (annagav) [inline ecommerce/api.py]: This Identifies if a user has paid for a particular course_run. For each paid course_run you have 2 exam attempts. So to schedule a third attempt you would need to pay.

File: ecommerce/api.py (hunk 1 above)
Comment (annagav) [inline ecommerce/api.py]: I this already tests that https://github.com/mitodl/micromasters/blob/8ee1a36d937e769abbde0fc45db50b538add37b3/ecommerce/api_test.py#L104

File: ecommerce/api_test.py [hunk 3]
```diff
@@ -216,6 +216,19 @@ def test_already_purchased(self):
 
         assert ex.exception.args[0] == 'Course run {} is already purchased'.format(course_run.edx_course_key)
 
+    @patch('ecommerce.api.has_to_pay_for_exam', return_value=True)
```
Comment (amir-qayyum-khan) [inline ecommerce/api_test.py]: please add a test where `has_to_pay_for_exam is False`

Comment (amir-qayyum-khan) [inline ecommerce/api_test.py]: add assert `ValidationError`

**Comment by annagav** [comment]
@amir-qayyum-khan, I addressed your comment

**Comment by amir-qayyum-khan** [comment]
screen shot 2017-08-21 at 4 55 03 am

**Pull request merged**